{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "9df29ca7",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import glob\n",
    "import subprocess\n",
    "import os\n",
    "import time\n",
    "import psutil\n",
    "import time\n",
    "from typing import NamedTuple"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "75e15a8d",
   "metadata": {},
   "source": [
    "## Notebook Utils"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a11d870",
   "metadata": {},
   "outputs": [],
   "source": [
    "class Results(NamedTuple):\n",
    "    \"\"\"Container for storing TVC or haplotype caller results.\"\"\"\n",
    "    sample_id: str\n",
    "    runtime_minutes: float\n",
    "    memory_mb: float\n",
    "    cores_used: int\n",
    "    vcf_path: str"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "714bbc17",
   "metadata": {},
   "outputs": [],
   "source": [
    "class BenchmarkResult(NamedTuple):\n",
    "    \"\"\"Container for storing benchmarking results.\"\"\"\n",
    "    sample_id: str\n",
    "    snp_precision: float\n",
    "    snp_recall: float\n",
    "    snp_f1: float\n",
    "    indel_precision: float\n",
    "    indel_recall: float\n",
    "    indel_f1: float"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "718b019d",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "# VCF and Benchmark regions from https://www.nist.gov/programs-projects/genome-bottle version v4.2.1\n",
    "BENCHMARK_VCF = \"HG001_GRCh38_1_22_v4.2.1_benchmark.vcf.gz\"\n",
    "BENCHMARK_REGIONS = \"HG001_GRCh38_1_22_v4.2.1_benchmark.bed\"\n",
    "REFERENCE_GENOME = \"hg38.fa\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7357b0e4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Specify the number of cores, 90 is recommended. \n",
    "CORES = 90"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9a3f8288",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_TVC(input_bam, input_ref, cores):\n",
    "    \"\"\"\n",
    "    Run TVC variant caller and monitor resource usage across all threads/processes.\n",
    "    \n",
    "    :param input_bam: Path to input BAM file.\n",
    "    :param input_ref: Path to reference genome file.\n",
    "    :param cores: Number of CPU cores to use.\n",
    "    :return: TVCResult containing runtime and memory usage information.\n",
    "    \"\"\"\n",
    "    output_vcf = input_bam.replace(\".bam\", \".tvc.vcf\")\n",
    "    cmd = f\"target/debug/tvc -t {cores} {input_ref} {input_bam} {output_vcf}\"\n",
    "\n",
    "    start_time = time.time()\n",
    "    proc = subprocess.Popen(cmd, shell=True)\n",
    "    ps_proc = psutil.Process(proc.pid)\n",
    "\n",
    "    peak_memory = 0\n",
    "    while proc.poll() is None:\n",
    "        try:\n",
    "            # Include child processes’ memory\n",
    "            mem_total = ps_proc.memory_info().rss\n",
    "            for child in ps_proc.children(recursive=True):\n",
    "                mem_total += child.memory_info().rss\n",
    "            peak_memory = max(peak_memory, mem_total)\n",
    "        except psutil.NoSuchProcess:\n",
    "            break\n",
    "        time.sleep(0.05)\n",
    "\n",
    "    end_time = time.time()\n",
    "    elapsed_minutes = (end_time - start_time) / 60\n",
    "\n",
    "    result = Results(\n",
    "        sample_id=input_bam.split(\".\")[0],\n",
    "        runtime_minutes=elapsed_minutes,\n",
    "        memory_mb=peak_memory / (1024 ** 2),\n",
    "        vcf_path=output_vcf,\n",
    "        cores_used=cores\n",
    "    )\n",
    "    return result\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1d46c999",
   "metadata": {},
   "outputs": [],
   "source": [
    "def run_hap_caller(input_bam, input_ref, cores):\n",
    "    \"\"\"\n",
    "    Run haplotype caller variant caller and monitor resource usage across all threads/processes.\n",
    "    \n",
    "    :param input_bam: Path to input BAM file.\n",
    "    :param input_ref: Path to reference genome file.\n",
    "    :param cores: Number of CPU cores to use.\n",
    "    :return: Results containing runtime and memory usage information.\n",
    "    \"\"\"\n",
    "    output_vcf = input_bam.replace(\".bam\", \".hap.vcf\")\n",
    "    cmd = f\"gatk HaplotypeCaller -t {cores} -R {input_ref} -I {input_bam} -O {output_vcf}\"\n",
    "\n",
    "    start_time = time.time()\n",
    "    proc = subprocess.Popen(cmd, shell=True)\n",
    "    ps_proc = psutil.Process(proc.pid)\n",
    "\n",
    "    peak_memory = 0\n",
    "    while proc.poll() is None:\n",
    "        try:\n",
    "            # Include child processes’ memory\n",
    "            mem_total = ps_proc.memory_info().rss\n",
    "            for child in ps_proc.children(recursive=True):\n",
    "                mem_total += child.memory_info().rss\n",
    "            peak_memory = max(peak_memory, mem_total)\n",
    "        except psutil.NoSuchProcess:\n",
    "            break\n",
    "        time.sleep(0.05)\n",
    "\n",
    "    end_time = time.time()\n",
    "    elapsed_minutes = (end_time - start_time) / 60\n",
    "\n",
    "    result = Results(\n",
    "        sample_id=input_bam.split(\".\")[0],\n",
    "        runtime_minutes=elapsed_minutes,\n",
    "        memory_mb=peak_memory / (1024 ** 2),\n",
    "        vcf_path=output_vcf,\n",
    "        cores_used=cores\n",
    "    )\n",
    "    return result"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "3671d036",
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_accuracy_metrics(benchmark_vcf, test_vcf, regions_bed, reference_genome):\n",
    "    \"\"\"\n",
    "    Extract accuracy metrics using hap.py Docker container.\n",
    "    :param benchmark_vcf: Path to benchmark VCF file.\n",
    "    :param test_vcf: Path to test VCF file.\n",
    "    :param regions_bed: Path to BED file defining regions of interest.\n",
    "    :param reference_genome: Path to reference genome file.\n",
    "    :return: BenchmarkResult containing precision, recall, and F1 scores for SNPs and INDELs.\n",
    "    \"\"\"\n",
    "    random_suffix = str(os.getpid())\n",
    "    docker_image = \"quay.io/biocontainers/hap.py:0.3.14--py27h5c5a3ab_0\"\n",
    "    cmd = f\"docker run --rm -v {os.getcwd()}:/data {docker_image} hap.py /data/{benchmark_vcf} /data/{test_vcf} -r /data/{reference_genome} -f /data/{regions_bed} -o /data/temp_happy_output.{random_suffix} --threads 90\"\n",
    "    subprocess.run(cmd, shell=True, stderr=subprocess.DEVNULL, stdout=subprocess.DEVNULL)\n",
    "    metrics_file = f'temp_happy_output.{random_suffix}.summary.csv'\n",
    "\n",
    "    df = pd.read_csv(metrics_file)\n",
    "    for row in df.itertuples():\n",
    "        if row.Filter != 'PASS':\n",
    "            continue\n",
    "        if row.Type == 'SNP':\n",
    "            snp_precision = row._12\n",
    "            snp_recall = row._11\n",
    "            snp_f1 = row._14\n",
    "        elif row.Type == 'INDEL':\n",
    "            indel_precision = row._12\n",
    "            indel_recall = row._11\n",
    "            indel_f1 = row._14\n",
    "\n",
    "    os.remove(metrics_file)\n",
    "\n",
    "    \n",
    "    result = BenchmarkResult(\n",
    "        sample_id=test_vcf.split(\".\")[0],\n",
    "        snp_precision=snp_precision,\n",
    "        snp_recall=snp_recall,\n",
    "        snp_f1=snp_f1,\n",
    "        indel_precision=indel_precision,\n",
    "        indel_recall=indel_recall,\n",
    "        indel_f1=indel_f1\n",
    "    )\n",
    "    return result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0edec7c6",
   "metadata": {},
   "source": [
    "## Analysis on Taps and Unconverted"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b096a049",
   "metadata": {},
   "outputs": [],
   "source": [
    "input_bams = glob.glob(\"*.deduped.bam\")\n",
    "tvc_results = []\n",
    "accuracy_results = []\n",
    "for bam in input_bams:\n",
    "    tvc_result = run_TVC(bam, REFERENCE_GENOME, cores=90)\n",
    "    accuracy_results.append(extract_accuracy_metrics(BENCHMARK_VCF, tvc_result.vcf_path, BENCHMARK_REGIONS, REFERENCE_GENOME))\n",
    "    tvc_results.append(tvc_result)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "385e605e",
   "metadata": {},
   "outputs": [],
   "source": [
    "tvc_df = pd.DataFrame(tvc_results)\n",
    "tvc_df = tvc_df.set_index('sample_id')\n",
    "accuracy_df = pd.DataFrame(accuracy_results)\n",
    "accuracy_df = accuracy_df.set_index('sample_id')\n",
    "\n",
    "final_df = pd.concat([tvc_df, accuracy_df], axis=1)\n",
    "final_df.to_csv(\"tvc_benchmarking_results.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0957dfb8",
   "metadata": {},
   "source": [
    "## Results on demo data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e36fa724",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/tmp/ipykernel_1467351/2730702165.py:13: FutureWarning: Calling int on a single element Series is deprecated and will raise a TypeError in the future. Use int(ser.iloc[0]) instead\n",
      "  depth_of_coverage = int(wgs_data_csv['MEDIAN_COVERAGE'])\n"
     ]
    }
   ],
   "source": [
    "demo_tvc_results = run_TVC(\"999999.bam\", REFERENCE_GENOME, cores=90)\n",
    "demo_happy_results = extract_accuracy_metrics(BENCHMARK_VCF, demo_tvc_results.vcf_path, BENCHMARK_REGIONS, REFERENCE_GENOME)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ee710a5c",
   "metadata": {},
   "outputs": [],
   "source": [
    "demo_tvc_df = pd.DataFrame([demo_tvc_results])\n",
    "demo_happy_df = pd.DataFrame([demo_happy_results])\n",
    "\n",
    "demo_final_df = pd.concat([demo_tvc_df, demo_happy_df], axis=1)\n",
    "demo_final_df.to_csv(\"demo_tvc_benchmarking_results.tsv\", sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "92a30ab4",
   "metadata": {},
   "source": [
    "## Below are results on the demo set"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3077db7e",
   "metadata": {},
   "outputs": [],
   "source": [
    "demo_final_df"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "83aa44f3",
   "metadata": {},
   "source": [
    "## Run haplotype caller for comparison\n",
    "### Make sure GATK is intalled first...\n",
    "Intallation instructions can be found here: \n",
    "\n",
    "https://gatk.broadinstitute.org/hc/en-us/articles/360041320571--How-to-Install-all-software-packages-required-to-follow-the-GATK-Best-Practices"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "60bc1e51",
   "metadata": {},
   "outputs": [],
   "source": [
    "hap_caller_results = []\n",
    "hap_caller_accuracy_results = []\n",
    "for bam in input_bams:\n",
    "    hap_caller_result = run_hap_caller(bam, REFERENCE_GENOME, cores=90)\n",
    "    hap_caller_accuracy_results.append(extract_accuracy_metrics(BENCHMARK_VCF, hap_caller_result.vcf_path, BENCHMARK_REGIONS, REFERENCE_GENOME))\n",
    "    hap_caller_results.append(hap_caller_result)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "014b72c7",
   "metadata": {},
   "outputs": [],
   "source": [
    "hap_caller_df = pd.DataFrame(hap_caller_results)\n",
    "hap_caller_df = hap_caller_df.set_index('sample_id')\n",
    "hap_caller_accuracy_df = pd.DataFrame(hap_caller_accuracy_results)\n",
    "hap_caller_accuracy_df = hap_caller_accuracy_df.set_index('sample_id')\n",
    "\n",
    "final_df = pd.concat([hap_caller_df, hap_caller_accuracy_df], axis=1)\n",
    "final_df.to_csv(\"hap_caller_benchmarking_results.tsv\", sep=\"\\t\")"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.13.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
